New Emailing Settings

Abdallah-Abouelatta opened the pull request:
Implementing customizable email settings that allow users to select whether they wish to receive confirmation emails or not.
By default, the 'upload_confirm_emails' checkbox, which allows users to opt-in to receive email notifications for each file upload they make, is set to 'False'. However, the 'attestation_emails' checkbox, which allows users to opt-in to receive email notifications for each task attestation, is set to 'True' by default.

Abdallah-Abouelatta committed:
UserChangeForm customised

'attestationEmails' and 'uploadConfirmEmails' checkboxes added to the UserChangeForm.

Abdallah-Abouelatta committed:
get and save checkboxes values

to get the values of the 'attestationEmails' and 'uploadConfirmEmails' checkboxes of the UserChangeForm

Abdallah-Abouelatta committed:
create new two fields win the 'User' class

'attestationEmails' and 'uploadConfirmEmails' have been added to the class User

Abdallah-Abouelatta committed:
Send submission email if statement customised 

to check if the user has selected the checkbox on the 'Change Account' page to confirm their willingness to receive an email fo each file upload ('False' by default)

Abdallah-Abouelatta committed:
send attestation confirmation email customised 

to check if the user has selected the checkbox on the 'Change Account' page to confirm their willingness to receive attestation emails ('True' by default)

physikerwelt commented on src/accounts/forms.py:
```diff
@@ -104,11 +104,13 @@ class UserChangeForm(forms.ModelForm):
     # overriding modelfields to ensure required fields are provided
     first_name = forms.CharField(max_length = 30, required=True)
     last_name = forms.CharField(max_length = 30, required=True)
+    attestationEmails = forms.BooleanField(help_text="recive Email for each task attestation.", required=False)
```
maybe use _ instead of camal case

ifrh replied:
`attestationEmails `=> `attestation_emails ` ?

physikerwelt commented on src/accounts/forms.py:
```diff
@@ -104,11 +104,13 @@ class UserChangeForm(forms.ModelForm):
     first_name = forms.CharField(max_length = 30, required=True)
     last_name = forms.CharField(max_length = 30, required=True)
+    attestationEmails = forms.BooleanField(help_text="recive Email for each task attestation.", required=False)
+    uploadConfirmEmails = forms.BooleanField(help_text="recive submission Email for each upload.", required=False)
```
also here _

ifrh replied:
`uploadConfirmEmails `=> `upload_confirm_emails` :-)

physikerwelt commented on src/accounts/views.py:
```diff
@@ -69,6 +69,29 @@ def change(request):
+            #to get the value of the 'attestationEmails' checkbox.
+            attestationEmails = request.POST.get('attestationEmails')
+            #to change the values of the value of the checkbox to 'True' or 'False'
+            is_checked = bool(attestationEmails)
```
this is never used

physikerwelt replied:
the comment should relate to line 73-88. I would guess

            user = form.save(commit=False)
            attestationEmails = form.cleaned_data['attestationEmails']
            uploadConfirmEmails = form.cleaned_data['uploadConfirmEmails']
            user.attestationEmails = attestationEmails
            user.uploadConfirmEmails = uploadConfirmEmails
            user.save()

does the job

physikerwelt replied:
@Abdallah-Abouelatta I mean that only the lines quoted above should remain; the rest assigns a variable that is never used and can be deleted.

physikerwelt reviewed (changes requested):
Looks good in general, some suggestions inline.

Abdallah-Abouelatta committed:
UserChangeForm new customisation 

'attestation_emails' and 'upload_confirm_emails' checkboxes added to the UserChangeForm.

ifrh reviewed (commented):
commit : https://github.com/KITPraktomatTeam/Praktomat/pull/362/commits/1dc36c765753670a2ed39761d296812226a326eb 
Looks good to me :-)

Abdallah-Abouelatta committed:
get and save checkboxes values 

to get the values of the 'attestation_emails' and 'upload_confirm_emails' checkboxes of the UserChangeForm

Abdallah-Abouelatta committed:
create new two fields win the 'User' class 

'attestation_emails' and 'upload_confirm_emails' have been added to the class User

ifrh reviewed (commented):
commit 79c508c30306344b8a2a287a105cf61ec71a41f3
looks good to me :-)

Abdallah-Abouelatta committed:
Send submission email if statement customised

to check if the user has selected the checkbox on the 'Change Account' page to confirm their willingness to receive an email for each file upload ('False' by default)

Abdallah-Abouelatta committed:
send attestation confirmation email customised 

to check if the user has selected the checkbox on the 'Change Account' page to confirm their willingness to receive attestation_emails ('True' by default)

Abdallah-Abouelatta committed:
saving the user model to the database. 

The commit=False in the form.save(commit=False) is used to prevent the form from saving the object to the database immediately.

physikerwelt commented on src/accounts/models.py:
```diff
@@ -42,6 +42,10 @@ class User(BasicUser):
     user_text=models.CharField(null=True, blank=True, max_length=500, help_text = _("Custom text which will be shown to this student."))
     accepted_disclaimer=models.BooleanField(default=False, help_text="Whether the user accepted the disclaimer.")
+    #'attestation_emails' to check whether the user wishes to receive an email notification for each attestation.
+    attestation_emails = models.BooleanField(default=True, help_text="recive confirmation Email for each task has been attested.")
```
I would update the English spelling of the help text to:
> Receive a confirmation email for each task that has been attested.

physikerwelt commented on src/accounts/models.py:
```diff
@@ -42,6 +42,10 @@ class User(BasicUser):
+    #'attestation_emails' to check whether the user wishes to receive an email notification for each attestation.
+    attestation_emails = models.BooleanField(default=True, help_text="recive confirmation Email for each task has been attested.")
+    #'upload_confirm_emails' to check whether the user wishes to receive an email notification for each upload they make.
+    upload_confirm_emails = models.BooleanField(default=False, help_text="recive confirmation Email for each file has been uploaded.")
```
see above

physikerwelt replied:
As you changed the default to false, this is a breaking change that should be reflected in the pull request message and maybe also in the changelog (if this exists for this project)

physikerwelt replied:
I guess there is no changelog. So putting it prominently in the PR description is sufficient.

physikerwelt commented on src/attestation/models.py:
```diff
@@ -65,7 +65,9 @@ def publish(self, request, by):
         headers = {'Reply-To': ', '.join(reply_to)} if reply_to else None
         email = EmailMessage(subject, body, None, (email,), headers = headers)
-        email.send()
+        #to check if the user has selected the checkbox on the 'Change Account' page to confirm their willingness to receive attestation emails ('True' by default)
```
Break this long line into two.

physikerwelt commented on src/solutions/views.py:
```diff
@@ -168,7 +168,8 @@ def solution_list(request, task_id, user_id=None):
 
                 else: #we are sending unsigned email
-                    if solution.author.email:
+                    #one of the checks is: if the user has selected the checkbox on the 'Change Account' page to confirm their willingness to receive an email fo each file upload ('False' by default)
```
Break this long line into two.

physikerwelt reviewed (changes requested):
Looks much better now, can you also fix the typo in the pull request title.

Abdallah-Abouelatta committed:
English spelling of the help text updated

'upload_confirm_emails' and 'attestation_emails' help text updated

Abdallah-Abouelatta committed:
comment modified

A line that was previously too long has been broken into two separate lines for better readability.

Abdallah-Abouelatta committed:
comment modified

A line that was previously too long has been broken into two separate lines for better readability.

physikerwelt reviewed (approved):
Thank you. I think this is good to merge.

Abdallah-Abouelatta committed:
The help text in 'UserChangeForm' updated 

'upload_confirm_emails' and 'attestation_emails' help text updated

kloenk reviewed (approved):
Looks reasonable

ifrh reviewed (changes requested):
Perhaps change 

https://github.com/KITPraktomatTeam/Praktomat/blob/49ab69f77bc55510540a388feea75ca612a4869c/src/solutions/views.py#L167-L168

to become like that:

```
 if solution.author.email and User.attestation_emails:
      message.send() 
```

Abdallah-Abouelatta committed:
if statement modified

'and User.upload_confirm_emails' added to the if-statement of sending the upload confirmation email.

ifrh reviewed (approved):
looks good to me.

ifrh commented on src/solutions/views.py:
```diff
@@ -168,7 +168,9 @@ def solution_list(request, task_id, user_id=None):
                          message.send() # any PY2-PY3 problem in here ?
```
you updated the if-clause :-) 

(how I can now obsolete my both comments ? )
